# mira/cli: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [FastQC](#fastqc) - Raw read QC
- [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipelilsne
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
- [Reads Prepped for IRMA](#reads-prepped-for-irma) - The subsampled and trimmed fastq inputs that IRMA will use for assembly
- [IRMA](#irma-outputs) - A folder containing all of the outs made by IRMA
- [DAIS-ribosome](dais-ribosome-outputs) - Aggregate insertion, deletion and seqeunce file for the input data
- [prepareIRMAjson](#prepare-IRMA-json) - Collected results from IRMA and DAIS-Ribosome in json files
- [statichtml](#static-html) - Creatge html, excel files and amended consensus fasta files
- [parquetmaker`](#parquet-maker) - Convert into parquet files

### FastQC

<details markdown="1">
<summary>Output files</summary>

- `fastqc/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

:::note
The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.
:::

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

### Reads Prepped for IRMA

<details markdown="1">
<summary>Output files</summary>

- `reads_prepped_for_IRMA:`
  - Subsampled read in fastq file
  - trimmed reads in fastq files for those experiment taypes that require trimming

</details>

### IRMA

<details markdown="1">
<summary>Output folder with sample name</summary>

IRMA output directory structre (only showing A_MP)

- `amended_consensus/`
  - Mixture_Example_7.fa - amended consensus
  - Mixture_Example_7.a2m - Optional amended global alignment to profile HMM
  - Mixture_Example_7.pad.fa - Optional N-padded cnosensus for amplicon dropouts.

- `figures/`
  - A_MP-coverageDiagram.pdf - Shows coverage and variant calls
  - A_MP-heuristics.pdf - Heuristic graphs for A_MP
  - A_MP-EXPENRD.pdf - A_MP variant phrasing using 'experimental enrichment' distances
  - A_MP-JACCARD.pdf - A_MP variant phasing using modified Jaccard distances
  - A_MP-MUTUALD.pdf - A_MP variant phasing using mutual association distances
  - A_MP-NJOINTP.pdf - A_MP variant phasing using normal;ized joint probability distances
  - READ_PERCENTAGES.pdf - Break down or reads assembled

- `intermediate/`
  - `0-ITERATIVE-REFERENCES/`
  - R0-A_MP.ref
  - R1-A_MP.ref
  - R2-A_MP.ref
  - `1-MATCH_BLAT/`
  - R1-tar.gz
  - R2-tar.gz
  - R3-tar.gz
  - `2-SORT_BLAT/`
  - R1.tar.gz
  - R1.txt
  - R2.tar.gz
  - R2.txt
  - `3-ALIGN_SAM/`
  - storedCounts.tar.gz
  - `4-ASSEMBLE_SSW/`
  -

</details>

### DAIS-ribosome

Aggregate insertion, deletion and seqeunce file for the input data
<details markdown="1">
<summary>Output files</summary>

- `reads_prepped_for_IRMA/`
  - Subsampled read in fastq file
  - trimmed reads in fastq files for those experiment taypes that require trimming

</details>

### prepareIRMAjson

Collected results from IRMA and DAIS-Ribosome in json files
<details markdown="1">
<summary>Output files</summary>

- `reads_prepped_for_IRMA/`
  - Subsampled read in fastq file
  - trimmed reads in fastq files for those experiment taypes that require trimming

</details>

### Statichtml

Creatge html, excel files and amended consensus fasta files
<details markdown="1">
<summary>Output files</summary>

- `reads_prepped_for_IRMA/`
  - Subsampled read in fastq file
  - trimmed reads in fastq files for those experiment taypes that require trimming

</details>

### parquetmaker

Convert into parquet files
<details markdown="1">
<summary>Output files</summary>

- `reads_prepped_for_IRMA/`
  - Subsampled read in fastq file
  - trimmed reads in fastq files for those experiment taypes that require trimming

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
